torch/csrc/jit/codegen/cuda/lower_shift.cpp

```diff
@@ -0,0 +1,684 @@
+    auto expr = *it;
+    if (!expr->outputs()[0]->isA<TensorView>()) {
+      continue;
+    }
+
+    propagateRootAxisInfo(expr);
+  }
+
+  // Propagates halo information from root axes down to leaf axes
+  for (auto tv : tvs) {
+    build(tv->domain());
+  }
+
+  // Note that validation requires consumer halo info
+  for (auto tv : tvs) {
+    validate(tv);
+  }
+}
+
+void HaloInfo::propagateRootAxisInfo(Expr* expr) {
+  for (auto output : expr->outputs()) {
+    auto out_tv = dynamic_cast<TensorView*>(output);
+    if (out_tv == nullptr) {
+      continue;
+    }
+    for (auto input : expr->inputs()) {
+      auto in_tv = dynamic_cast<TensorView*>(input);
+      if (in_tv == nullptr) {
+        continue;
+      }
+      propagateRootAxisInfo(in_tv, out_tv, expr);
+    }
+  }
+}
+
+void HaloInfo::propagateRootAxisInfo(
+    TensorView* producer,
+    TensorView* consumer,
+    Expr* expr) {
+  // Do not add halo to input tensors
+  if (producer->isFusionInput()) {
+    return;
+  }
+
+  auto c2p = PairwiseRootDomainMap(producer, consumer)
+                 .mapConsumerToProducer(consumer->domain(), producer->domain());
+
+  const auto& c_root = consumer->getRootDomain();
+
+  for (size_t i = 0; i < c_root.size(); ++i) {
+    auto c_id = c_root[i];
+    auto it = c2p.find(c_id);
+    if (it == c2p.end()) {
+      // nothing to propagate
+      continue;
+    }
+
+    // propagate root-axis halo info from c_id to p_id
+
+    auto p_id = it->second;
+
+    auto p_info = getRootAxisInfo(p_id);
+    const auto c_info = getRootAxisInfo(c_id);
+
+    // If the defining expression is shift, adjust the producer halo
+    // width based on the shift offset. If the shift offset is
+    // positive, create halo at offset zero of the producer axis so
+    // that the consumer can safely access the producer. If the offset
+    // is negative, halo is created at the other end of the axis.
+    // If the expr is not shift, just merge the consumer halo info
+    // to the producer halo info so that the producer halo can be the
+    // maximum of all its consumers.
+    if (auto shift_op = dynamic_cast<ShiftOp*>(expr)) {
+      const int offset = shift_op->offset(i);
+      if (offset == 0) {
+        p_info.merge(c_info);
+      } else {
+        int pos = (offset > 0) ? 0 : 1;
+        p_info.merge(pos, c_info.width(pos) + std::abs(offset));
+      }
+    } else {
+      p_info.merge(c_info);
+    }
+    setRootAxisInfo(p_id, p_info);
+  }
+}
+
+// Propagate extent information from root axes to descendants
+void HaloInfo::build(TensorDomain* td) {
+  auto gpu_lower = GpuLower::current();
+
+  for (auto root_axis : td->getRootDomain()) {
+    const auto& halo_info = getRootAxisInfo(root_axis);
+    auto halo_width = halo_info.width();
+
+    // When root_axis is used as part of a reference tensor, update is
+    // called multiple times, which should be safe. Just for safety,
+    // make sure it uses the same halo width.
```

Comment:
Thanks, that's a good point. Multiple calls should never occur anymore, so this can be more restrictive, i.e., no existing mapping should be found. I'll update the code and comment.